src/core/backbone_router/local.cpp

```diff
@@ -238,6 +241,92 @@ void Local::UpdateBackboneRouterPrimary(Leader::State aState, const BackboneRout
     return;
 }
 
+otError Local::GetDomainPrefix(NetworkData::OnMeshPrefixConfig &aConfig)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(mDomainPrefixConfig.mPrefix.mLength > 0, error = OT_ERROR_NOT_FOUND);
+
+    aConfig = mDomainPrefixConfig;
+
+exit:
+    return error;
+}
+
+otError Local::RemoveDomainPrefix(const otIp6Prefix &aPrefix)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aPrefix.mLength > 0, error = OT_ERROR_INVALID_ARGS);
+
+    VerifyOrExit(mDomainPrefixConfig.mPrefix.mLength == aPrefix.mLength, error = OT_ERROR_NOT_FOUND);
+
+    VerifyOrExit(otIp6PrefixMatch(&mDomainPrefixConfig.mPrefix.mPrefix, &aPrefix.mPrefix) >=
+                     mDomainPrefixConfig.mPrefix.mLength,
+                 error = OT_ERROR_NOT_FOUND);
+
+    mDomainPrefixConfig.mPrefix.mLength = 0;
+
+    if (IsEnabled())
+    {
+        SuccessOrExit(error =
+                          Get<NetworkData::Local>().RemoveOnMeshPrefix(aPrefix.mPrefix.mFields.m8, aPrefix.mLength));
+        Get<NetworkData::Notifier>().HandleServerDataUpdated();
+    }
+
+exit:
+    return error;
+}
+
+otError Local::SetDomainPrefix(const NetworkData::OnMeshPrefixConfig &aConfig)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aConfig.mPrefix.mLength > 0, error = OT_ERROR_INVALID_ARGS);
+
+    if (IsEnabled())
+    {
+        Get<NetworkData::Local>().RemoveOnMeshPrefix(mDomainPrefixConfig.mPrefix.mPrefix.mFields.m8,
+                                                     mDomainPrefixConfig.mPrefix.mLength);
+    }
+
+    mDomainPrefixConfig = aConfig;
```

Comment:
```
> prefix
prefix
2001:dead:beef:cafe::/64 paros med
fd00:7d03:7d03:7d03::/64 prosD med
------------------------------------
fd00:7d03:7d03:7d03::/64 prosD med [ot::BackboneRouter::Local]
Done
> ipaddr mleid
ipaddr mleid
fdde:ad00:beef:0:63b4:ac98:130:649c
Done
> prefix add fdde:ad00:beef:0::/64 prosD med
prefix add fdde:ad00:beef:0::/64 prosD med
Error 7: InvalidArgs
> prefix
prefix
2001:dead:beef:cafe::/64 paros med
------------------------------------
fdde:ad00:beef:0::/64 prosD med [ot::BackboneRouter::Local]
Done
```

`prefix add <mleid> prosD med` can remove the domain prefix in local data and overwrite `mDomainPrefixConfig`, but eventually failed to add to local data. Should never allow using `mleid` as domain prefix?

Reply:
Yes, Domain Prefix must not share the same prefix as Mesh Local. Should add a check for this.

Reply:
If so, we should also add check for modifying `MLEID`.

Reply:
Furthermore, is it possible that a `BBR` capable node could join a Thread network with `mleid` equals to its previously configured domain prefix?  Should we clear `mDomainPrefixConfig` when node attaches in such case?

Reply:
It might be simpler to not allow enabling BBR service if Domain Prefix matches mesh-local prefix.

Note that it's not just the ML-EID, it's anything that matches mesh-local prefix.

Reply:
Thanks for clarification. No enabling BBR service when `ML` is a prefix of domain prefix seems simplest to implement.

Reply:
I don't think we can reject an operational dataset - only the Leader can do that.

How about we simply disable BBR service if the new active operational dataset has a conflicting ML prefix?

Reply:
I found that it is possible to enable BBR service without a domain prefix. If a BBR service without domain prefix is valid and useful, maybe we should just turnoff domain prefix when it conflicts with ML prefix and leave BBR service on? Seems better to have minimal surprising impact.

Reply:
Sorry when I was saying `reject such pending dataset at the first beginning`, I mean rejection by leader, not allowing such dataset change.

I am fine with the simple option to disable BBR service if ML prefix conflicts when applying operational dataset.

Reply:
> Sorry when I was saying reject such pending dataset at the first beginning, I mean rejection by leader, not allowing such dataset change.

I think it is not easy to implement this correctly. When we configure active/pending dataset on the Leader, the `SVR_DATA.ntf` message from the commissioner could be already transmitting in the air, but the Leader haven't received it. The Leader would need extra logic to reject this `SVR_DATA.ntf`, and cleanup current Network Data if necessary.

Reply:
We have to assume that the Leader may be a Thread 1.1 device.

Reply:
Thanks for the valuable discussions here. I would summarize below scenarios if only speaking of possible coincidence:

Note: below wording `on-mesh prefix` indicates non-mesh local prefix for easier description.

1) For any Thread 1.1 or Thread 1.2 leader, would it be allowed to
     a) accept on mesh prefix registration via SRV_DATA.ntf which may conflicts with the one  in pending dataset?
      b) accept pending operation dataset which may have mesh local prefix the same as the other configured on mesh prefix which was not mesh local prefix from some authorized source via MGMT_PENDING_SET ?

2) For any Thread 1.1/1.2 server, when it joins the Thread Network, if the local on-mesh prefix configuration is the same as mesh local prefix of the network it joins, should it somehow notify upper layer and remove or suppress its SRV_DATA.ntf? 

3) Similar as 2) , suppose a Thread 1.1/1.2 servers migrates to a Network of which the pending dataset has the same mesh local prefix as the on-mesh prefix configuration of the server, what's the expected behavior?

Domain Prefix in Thread 1.2  belongs to the on-mesh prefixes, I think we may need to consider the common coincidence issue between on-mesh prefix  and mesh local prefix.

As for whether or not disable BBR Service when domain prefix conflicts with mesh local prefix, I think there is no clear definition in Thread Specification, it could be flexible for the implementation. We may keep BBR service if we think the multicast feature should be supported if possible.

Thoughts? @jwhui @simonlingoogle

Reply:
@librasungirl, thank you for summarizing the questions. I agree that handling all the situations properly is complex. I am fine if OpenThread allows such configuration and effectively says that configuring on-mesh/domain prefix the same as mesh-local prefix is a user/admin configuration problem. If this becomes an issue later, we can address it then.

Reply:
agree. thanks.